### DISCUSSION TITLE
Whitespace issue with custom Bard logic

### DISCUSSION THREAD

**Discussion by carstenjaksch** [statamic/cms#11304] [Q&A] [answered]
I use a custom loop to build my own Bard rendering logic. This allows me to render partials for parts of Bard text:

```antlers
    {{ foreach :array="bard|raw" }}
        {{ if value:type == 'heading' }}
            {{ switch(
                (value:attrs:level == '2') => { partial:typography/heading :heading="value:content:0:text" },
                (value:attrs:level == '3') => { partial:typography/heading as="h3" :heading="value:content:0:text" },
                (value:attrs:level == '4') => { partial:typography/heading as="h4" :heading="value:content:0:text" },
                () => { text }
            ) }}
        {{ elseif value:type == 'paragraph' }}
            <p
                {{ value:attrs:class ?= 'class="{value:attrs:class}"' }}
                {{ value:attrs:textAlign == 'left' ? '' : 'style="text-align: {value:attrs:textAlign}"' }}
            >
                {{ partial:typography/paragraph :content="value:content" }}
            </p>
        {{ elseif value:type == 'bulletList' }}
            <ul>
                {{ value:content }}
                    <li>
                        {{ partial:typography/paragraph :content="content:0:content" }}
                    </li>
                {{ /value:content }}
            </ul>
        {{ /if }}
    {{ /foreach }}
```

The paragraph partial:

```antlers
{{ content }}
    {{ if type == 'hardBreak' }}
        <br>
    {{ elseif marks }}
        {{ if marks:0:type == 'bold' }}
            <strong>{{ text }}</strong>
        {{ elseif marks:0:type == 'link' }}
            {{ href = marks:0:attrs:href|explode('::')|last }}
            {{ url = href|url }}

            <a
                href="{{ url ? url : href }}"
                {{ marks:0:attrs:rel ?= 'rel="{marks:0:rel}"' }}
                {{ marks:0:attrs:target ?= 'target="{marks:0:target}"' }}
                {{ marks:0:attrs:title ?= 'title="{marks:0:title}"' }}
                class="inline-block"
            >{{ text|antlers }}</a>
        {{ /if }}
    {{ else }}
        {{ text|antlers }}
    {{ /if }}
{{ /content }}
```

Works great so far, but I have a little nasty issue with punctuation that immediately follows a link in the paragraph partial. There is a whitespace between the two:

![Bildschirm­foto 2024-12-26 um 23 55 52](https://github.com/user-attachments/assets/06692c68-e35c-451b-8578-3d18d94598f8)

Because of the loop, the dot is placed on a new line in the HTML output and the browser renders a space.

I know there is a `spaceless` modifier that should solve the issue, but I cannot use it with a partial tag. Is there another way to get rid of that space?

**Comment by edalzell** [comment]
Why can't you use it with the spaceless modifier?

**Reply by carstenjaksch** [reply]
Thank you for your answer! Because no modifier works with a partial tag: `{{ partial:xyz | spaceless }}` or `dump` has no effect. Maybe I'm doing something wrong.

**Reply by edalzell** [reply, marked as answer]
`{{ { partia:...... } | spaceless }}`

**Reply by carstenjaksch** [reply]
That worked, thanks! I used a `regex_replace` modifier in this case.

Maybe these use-cases could be documented better somehow:

[Partial](https://statamic.dev/tags/partial)

Missing part about usage with modifiers. Or add that part [here](https://statamic.dev/modifiers) and link to it.

[regex_replace](https://statamic.dev/modifiers/regex_replace)

Does not have a regex example. Some things I found out:

1. Omit `/` at the start and end of a regex. But how to use regex options like `g` here?
2. Use `\1` instead of `$1` for the replacement.

**Reply by edalzell** [reply]
> Missing part about usage with modifiers. 

It has nothing to do with the partial tag, you can do that with any tag, really, or even hard coded data `{{ "foo " | spaceless }}`

**Reply by edalzell** [reply]
> Does not have a regex example.

PRs gladly accepted!

**Reply by carstenjaksch** [reply]
> It has nothing to do with the partial tag, you can do that with any tag, really, or even hard coded data `{{ "foo " | spaceless }}`

Yeah, but the pitfall with partials is that you need to wrap them in a separate curly brace to use modifiers. Maybe a bit confusing for beginners (like me).

> PRs gladly accepted!

Sure thing. However, I am a bit intimidated by my limited knowledge and English is not my native language.

I can't think of a good example and someone would have to (completely) rewrite my PR anyways. And I have no idea about the Blade part.

So many things that make even small PRs really hard for me. Especially with my perfectionism. 😄 

Just wanted to point out weak points in the docs, maybe someone else with better skills could pick these up. Although a separate issue would probably be a better place for that.

**Reply by edalzell** [reply]
> Yeah, but the pitfall with partials is that you need to wrap them in a separate curly brace to use modifiers.

Yup it's a sub-expression: https://statamic.dev/antlers#subexpressions

**Reply by edalzell** [reply]
> Although a separate issue would probably be a better place for that.

An issue on the `docs` repo would be great, thanks.
